File: routes/locations.py
from fastapi import APIRouter, Depends, HTTPException
from mongoengine import DoesNotExist, ValidationError, NotUniqueError
from typing import List
from schemas.locations import LocationSchema, LocationCreate
from models.models import User, Location as LocationModel, Fact, Tag
from utils.auth_utils import get_current_user
import logging
import traceback

router = APIRouter(prefix="/api/v3")

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

@router.get("/locations", response_model=List[LocationSchema])
async def get_user_locations(current_user: User = Depends(get_current_user)):
    try:
        logger.info(f"Fetching locations for user: {current_user.id}")
        user_locations = LocationModel.objects(user_id=current_user.id)
        logger.info(f"Retrieved {len(user_locations)} locations for user: {current_user.id}")
        return [LocationSchema(
            id=str(location.id),
            name=location.name,
            latitude=location.latitude,
            longitude=location.longitude,
            description=location.description,  
            user_id=str(current_user.id),
            created_at=location.created_at,
            tags=[tag.name for tag in location.tags]  # Add this line
        ) for location in user_locations]
    except Exception as e:
        logger.error(f"Error fetching locations for user {current_user.id}: {str(e)}")
        logger.error(traceback.format_exc())
        raise HTTPException(status_code=500, detail="An error occurred while fetching locations")

@router.post("/locations", response_model=LocationSchema)
async def create_location(location: LocationCreate, current_user: User = Depends(get_current_user)):
    try:
        logger.info(f"Creating new location for user: {current_user.id}")
        tag_objects = [Tag.objects.get(name=tag) for tag in location.tags]
        new_location = LocationModel(
            name=location.name,
            latitude=location.latitude,
            longitude=location.longitude,
            description=location.description,
            user_id=current_user,
            tags=tag_objects  # Add this line
        )
        new_location.save()

        # Create a corresponding Fact
        new_fact = Fact(
            description=new_location.description,
            location_id=new_location,
            user_id=current_user
        )
        new_fact.save()

        logger.info(f"Location created with ID: {new_location.id} for user: {current_user.id}")
        return LocationSchema(
            id=str(new_location.id),
            name=new_location.name,
            latitude=new_location.latitude,
            longitude=new_location.longitude,
            description=new_location.description,
            user_id=str(current_user.id),
            created_at=new_location.created_at,
            tags=[tag.name for tag in new_location.tags]  # Add this line
        )
    except NotUniqueError:
        logger.error(f"Unique constraint violation for user {current_user.id}")
        raise HTTPException(status_code=400, detail="Location with the same latitude and longitude already exists.")
    except ValidationError as e:
        logger.error(f"Validation error creating location for user {current_user.id}: {str(e)}")
        raise HTTPException(status_code=400, detail="Invalid data provided.")
    except Exception as e:
        logger.error(f"Unexpected error creating location for user {current_user.id}: {str(e)}")
        raise HTTPException(status_code=500, detail="An unexpected error occurred")

@router.put("/locations/{location_id}", response_model=LocationSchema)
async def update_location(location_id: str, location: LocationCreate, current_user: User = Depends(get_current_user)):
    logger.info(f"Updating location ID: {location_id} for user: {current_user.id}")
    try:
        existing_location = LocationModel.objects.get(id=location_id, user_id=current_user.id)
        existing_location.name = location.name
        existing_location.latitude = location.latitude
        existing_location.longitude = location.longitude
        existing_location.description = location.description
        existing_location.tags = [Tag.objects.get(name=tag) for tag in location.tags]  # Add this line
        existing_location.save()

        # Update corresponding Facts
        facts = Fact.objects(location_id=existing_location.id)
        for fact in facts:
            fact.description = existing_location.description
            fact.save()

        logger.info(f"Location updated with ID: {existing_location.id} for user: {current_user.id}")
        return LocationSchema(
            id=str(existing_location.id),
            name=existing_location.name,
            latitude=existing_location.latitude,
            longitude=existing_location.longitude,
            description=existing_location.description,
            user_id=str(current_user.id),
            created_at=existing_location.created_at,
            tags=[tag.name for tag in existing_location.tags]  # Add this line
        )
    except DoesNotExist:
        logger.warning(f"Location ID {location_id} not found for user: {current_user.id}")
        raise HTTPException(status_code=404, detail="Location not found.")
    except ValidationError as e:
        logger.error(f"Validation error updating location ID {location_id} for user {current_user.id}: {str(e)}")
        raise HTTPException(status_code=400, detail="Invalid data provided.")
    except Exception as e:
        logger.error(f"Unexpected error updating location ID {location_id} for user {current_user.id}: {str(e)}")
        raise HTTPException(status_code=500, detail="An unexpected error occurred")

@router.delete("/locations/{location_id}", status_code=204)
async def delete_location(location_id: str, current_user: User = Depends(get_current_user)):
    logger.info(f"Attempting to delete location ID: {location_id} for user: {current_user.id}")
    try:
        db_location = LocationModel.objects.get(id=location_id, user_id=str(current_user.id))
        db_location.delete()
        logger.info(f"Location ID {location_id} deleted successfully for user: {current_user.id}")
    except DoesNotExist:
        logger.warning(f"Location ID {location_id} not found for user: {current_user.id}")
        raise HTTPException(status_code=404, detail="Location not found.")

@router.get("/locations/count", response_model=int)
async def get_locations_count(current_user: User = Depends(get_current_user)):
    try:
        logger.info(f"Counting locations for user: {current_user.id}")
        count = LocationModel.objects(user_id=current_user.id).count()
        logger.info(f"User {current_user.id} has {count} locations")
        return count
    except Exception as e:
        logger.error(f"Error counting locations for user {current_user.id}: {str(e)}")
        logger.error(traceback.format_exc())
        raise HTTPException(status_code=500, detail="An error occurred while counting locations")

File: routes/locations.js
const express = require("express");
const mongoose = require("mongoose");
const { Location, Fact, Tag, User } = require("../models/models"); // Import Mongoose models
const { authenticateToken } = require("../utils/authenticate_token");
const logger = require("../utils/logger");

const router = express.Router();

// Fetch user locations
router.get("/locations", authenticateToken, async (req, res) => {
  const userId = req.user.id;
  logger.info(`Fetching locations for user: ${userId}`);

  try {
    const userLocations = await Location.find({ user_id: userId }).populate("tags");  // Change this line
    logger.info(`Retrieved ${userLocations.length} locations for user: ${userId}`);

    const response = userLocations.map((location) => ({
      id: location._id,
      name: location.name,
      latitude: location.latitude,
      longitude: location.longitude,
      description: location.description,
      //tags: location.tags,
    }));

    res.status(200).json(response);
  } catch (error) {
    logger.error(`Error fetching locations for user ${userId}: ${error.message}`);
    res.status(500).json({ message: "Error fetching locations" });
  }
});


// Create a new location
router.post("/locations", authenticateToken, async (req, res) => {
  const userId = req.user.id;
  const { name, latitude, longitude, description, tags } = req.body;

  logger.info(`Creating new location for user: ${userId}`);

  try {
    // Check if the user has already created a location with the same details
    const existingLocation = await Location.findOne({ user_id: userId, name, latitude, longitude, description });
    if (existingLocation) {
      logger.warn(`User ${userId} already has a location with the same details`);
      return res.status(400).json({ error: "You have already created a location with the same details." });
    }

    const tagObjects = await Tag.find({ name: { $in: tags } });

    const newLocation = new Location({
      name,
      latitude,
      longitude,
      description,
      user_id: userId,
      tags: tagObjects,
    });
    await newLocation.save();

    const newFact = new Fact({
      description,
      location_id: newLocation._id,
      user_id: userId,
    });
    await newFact.save();

    logger.info(`Location created with ID: ${newLocation._id} for user: ${userId}`);
    res.status(201).json({
      id: newLocation._id,
      name: newLocation.name,
      latitude: newLocation.latitude,
      longitude: newLocation.longitude,
      description: newLocation.description,
      tags: newLocation.tags,
    });
  } catch (error) {
    logger.error(`Error creating location for user ${userId}: ${error.message}`);
    res.status(500).json({ message: "Error creating location" });
  }
});

// Update a location
router.put("/locations/:locationId", authenticateToken, async (req, res) => {
  const userId = req.user.id;
  const { locationId } = req.params;
  const { name, latitude, longitude, description, tags } = req.body;

  logger.info(`Updating location ID: ${locationId} for user: ${userId}`);

  try {
    const existingLocation = await Location.findOne({ _id: locationId, user_id: userId }); // Ensure correct field name
    if (!existingLocation) {
      logger.warn(`Location ID ${locationId} not found for user: ${userId}`);
      return res.status(404).json({ error: "Location not found." });
    }

    const tagObjects = await Tag.find({ name: { $in: tags } });

    existingLocation.name = name;
    existingLocation.latitude = latitude;
    existingLocation.longitude = longitude;
    existingLocation.description = description;
    existingLocation.tags = tagObjects;

    await existingLocation.save();
    logger.info(`Location ID: ${locationId} updated successfully for user: ${userId}`);
    res.status(200).json({ message: "Location updated successfully" });
  } catch (error) {
    logger.error(`Error updating location ID ${locationId} for user ${userId}: ${error.message}`);
    res.status(500).json({ message: "Error updating location" });
  }
});

// Delete a location
router.delete("/locations/:locationId", authenticateToken, async (req, res) => {
  const userId = req.user.id;
  const { locationId } = req.params;

  logger.info(`Attempting to delete location ID: ${locationId} for user: ${userId}`);

  try {
    const location = await Location.findOneAndDelete({ _id: locationId, user_id: userId }); // Ensure correct field name
    if (!location) {
      logger.warn(`Location ID ${locationId} not found for user: ${userId}`);
      return res.status(404).json({ error: "Location not found." });
    }

    logger.info(`Location ID ${locationId} deleted successfully for user: ${userId}`);
    res.status(204).send();
  } catch (error) {
    logger.error(`Error deleting location ID ${locationId} for user ${userId}: ${error.message}`);
    res.status(500).json({ message: "Error deleting location" });
  }
});


// Count user locations
router.get("/locations/count", authenticateToken, async (req, res) => {
  const userId = req.user.id;

  logger.info(`Counting locations for user: ${userId}`);

  try {
    const count = await Location.countDocuments({ user_id: userId }); // Ensure correct field name
    res.status(200).json({ count });
  } catch (error) {
    logger.error(`Error counting locations for user ${userId}: ${error.message}`);
    res.status(500).json({ message: "Error counting locations" });
  }
});


// Fetch all locations for all users
router.get("/locations/all", authenticateToken, async (req, res) => {
  logger.info(`Fetching all locations for all users`);

  try {
    const allLocations = await Location.find().populate("tags");
    logger.info(`Retrieved ${allLocations.length} locations for all users`);

    const response = allLocations.map((location) => ({
      id: location._id,
      name: location.name,
      latitude: location.latitude,
      longitude: location.longitude,
      description: location.description,
      user_id: location.user_id,
      tags: location.tags,
    }));

    res.status(200).json(response);
  } catch (error) {
    logger.error(`Error fetching all locations: ${error.message}`);
    res.status(500).json({ message: "Error fetching all locations" });
  }
});

module.exports = router;
